### ISSUE TITLE
Optimal way to handle Bert attention mask with Flex Attention

### ISSUE THREAD

**Issue by conceptofmind** [pytorch-labs/attention-gym#208]
Hello,

I was wondering if there was a more optimal way to handle attention masking for BERT with Flex Attention.

A minimal reproduction:
```py
import torch
from transformers import AutoTokenizer, DataCollatorForLanguageModeling
from torch.nn.attention.flex_attention import flex_attention, create_block_mask

torch.nn.attention.flex_attention._FLEX_ATTENTION_DISABLE_COMPILE_DEBUG = True

device = "cuda"

tok = AutoTokenizer.from_pretrained("bert-base-uncased")
collator = DataCollatorForLanguageModeling(
    tokenizer=tok,
    mlm=True,
    mlm_probability=0.15,
)

examples = [
    tok("hello world", return_special_tokens_mask=True),
    tok("a much longer sentence for padding", return_special_tokens_mask=True),
]

batch = collator(examples)
input_ids = batch["input_ids"].to(device)
attention_mask = batch["attention_mask"].to(device).bool()

B, S = input_ids.shape
H, D = 2, 8

q = torch.randn(B, H, S, D, device=device)
k = torch.randn(B, H, S, D, device=device)
v = torch.randn(B, H, S, D, device=device)

def bert_mask_mod(b, h, q_idx, kv_idx):
    return attention_mask[b, q_idx] & attention_mask[b, kv_idx]

block_mask = create_block_mask(
    bert_mask_mod,
    B=B,
    H=H,
    Q_LEN=S,
    KV_LEN=S,
    device=device,
)

out = flex_attention(q, k, v, block_mask=block_mask)

print("input_ids:", input_ids.shape)
print("attention_mask:", attention_mask)
print("labels:", batch["labels"].shape)
print("out:", out.shape)
```
I had looked into using fa4 with cutedsl mask mod support following these articles:
https://pytorch.org/blog/flexattention-flashattention-4-fast-and-flexible/
https://research.colfax-intl.com/a-users-guide-to-flexattention-in-flash-attention-cute-dsl/

But was running into some issues.

Thank you,

Enrico

**Comment by drisspg** [comment]
Yeah there alre likely alot of things that can be done but the first thing I would ask is, do you know how Optimal your current implementation is? 

as well is see that your embedding dim is 8 Which is actually not supported today

**Comment by conceptofmind** [comment]
> Yeah there alre likely alot of things that can be done but the first thing I would ask is, do you know how Optimal your current implementation is?
> 
> as well is see that your embedding dim is 8 Which is actually not supported today

That was just an example dim haha. I can try to get a good benchmark vs varlen fa4. Currently was evaluating against sdpa where it was faster.

Pytorch nightly version of flex attention with flash 4 backend did not seem to work for our gh200 systems.

**Comment by drisspg** [comment]
hmm I can take a look at h200, I have spent more energy on blackwell then hopper, for all by dynamic shapes path though hopper should be equivalent. 

In general though you are curious if flex + MLM style masking could be sped up right ? that is the particular pattern you are trying to speed up?

**Comment by conceptofmind** [comment]
> hmm I can take a look at h200, I have spent more energy on blackwell then hopper, for all by dynamic shapes path though hopper should be equivalent.
> 
> In general though you are curious if flex + MLM style masking could be sped up right ? that is the particular pattern you are trying to speed up?

Correct, that is the particular pattern that I am trying to speed up. BERT encoder model with MLM-style masking. Right now I use flash 4 varlen attn with either padding and unpadding the inputs/outputs, or using a fully unpadded and packed data collator. Being able to plug in flex attention would simplify the code greatly.

**Comment by drisspg** [comment]
hmm so just to triple clarify my the returned masked from hf is the padding mask not what I think of as MLM. Which is chocolate chips (masked tokens) interleaved throughout a cookie (sentence) bidirectionally. You have support for the later in in FA4 varlen? are you also passing in a mask func directly there?

**Comment by conceptofmind** [comment]
> hmm so just to triple clarify my the returned masked from hf is the padding mask not what I think of as MLM. Which is chocolate chips (masked tokens) interleaved throughout a cookie (sentence) bidirectionally. You have support for the later in in FA4 varlen? are you also passing in a mask func directly there?

I currently have two versions.

One flash 4 varlen model that takes in attention masking:
```
        inputs: (batch, seqlen, ...) or (batch, seqlen)
        attention_mask: (batch, seqlen), bool / int, 1 means valid and 0 means not valid.
        labels: (batch, seqlen), int, labels
```
With MLM supplied from https://huggingface.co/docs/transformers/en/main_classes/data_collator#transformers.DataCollatorForLanguageModeling

Then I have a second flash 4 varlen version which takes:
```
        inputs: (total_nnz, ...)
        cu_seqlens: (batch + 1), the cumulative sequence lengths
        max_seqlen_in_batch: int
        labels: (total_nnz)
```
With a custom data collator that I built based on the collator above. This collator also ensures that the examples are divisible by 8 for chunked linear attention.

The first padded version may be slightly slower due to some additional slicing, but I can use the standard collator from HF.

The second version is slightly faster, but the collator is much more complex.

Both are bidirectional. In the first, I pass the attention mask. Currently, for pre-training a BERT model for MLM.

I can also show code if that is useful haha.

Being able to use flex attention with the MLM collator would be the simplest implementation.

**Comment by drisspg** [comment]
In general you would try and minimize attn_mask's size as much as possible. I and for fa4 indexing on the with the q_idx is faster on balckwell it is not as straightforward for hopper. 

But yeah I think roofline perf would be the integration we added to varlen. That being said we should be able to support for flex, any code would be helpful

**Comment by conceptofmind** [comment]
> In general you would try and minimize attn_mask's size as much as possible. I and for fa4 indexing on the with the q_idx is faster on balckwell it is not as straightforward for hopper.
> 
> But yeah I think roofline perf would be the integration we added to varlen. That being said we should be able to support for flex, any code would be helpful

I will clean up the code a little bit and then share. Thank you.

**Comment by conceptofmind** [comment]
```python
from typing import Optional

import torch
import torch.nn as nn
from flash_attn_interface import flash_attn_varlen_func
from quack.linear_cross_entropy import chunked_linear_cross_entropy
from quack.mlp import MLP
from quack.rmsnorm import QuackRMSNorm as RMSNorm
from quack.rotary import apply_rotary_emb


class PredictionHead(nn.Module):
    def __init__(
        self,
        dim: int,
        bias: bool = False,
    ):
        super().__init__()
        self.dense = nn.Linear(dim, dim, bias=bias)
        self.act = lambda x: torch.relu(x).square()
        self.norm = RMSNorm(dim)

    def forward(self, hidden_states: torch.Tensor) -> torch.Tensor:
        return self.norm(self.act(self.dense(hidden_states)))


class TokenEmbeddings(nn.Module):
    def __init__(self, num_classes, dim, pad_token_id=None):
        super().__init__()
        self.tok_embeddings = nn.Embedding(
            num_classes,
            dim,
            padding_idx=pad_token_id,
        )
        self.norm = RMSNorm(dim)
        nn.init.normal_(self.tok_embeddings.weight, std=0.02)

    def forward(self, input_ids: torch.LongTensor) -> torch.Tensor:
        x = self.tok_embeddings(input_ids)
        return self.norm(x)


class Attention(nn.Module):
    def __init__(self, dim, heads, dim_head, causal: bool = True):
        super().__init__()
        inner_dim = dim_head * heads
        self.heads = heads
        self.head_dim = dim_head
        self.causal = causal
        self.scale = dim_head**-0.5
        self.to_qkv = nn.Linear(dim, inner_dim * 3, bias=False)
        self.q_norm = RMSNorm(dim_head)
        self.k_norm = RMSNorm(dim_head)
        self.to_out = nn.Linear(inner_dim, dim, bias=False)

    def forward(
        self,
        hidden_states: torch.Tensor,
        cos: torch.Tensor,
        sin: torch.Tensor,
        cu_seqlens_q: Optional[torch.Tensor] = None,
        cu_seqlens_k: Optional[torch.Tensor] = None,
        max_seqlen_q: Optional[int] = None,
        max_seqlen_k: Optional[int] = None,
    ):
        total_nnz, dim = hidden_states.shape
        qkv = self.to_qkv(hidden_states)
        qkv = qkv.view(total_nnz, 3, self.heads, self.head_dim)
        q, k, v = qkv.unbind(dim=1)
        q = self.q_norm(q)
        k = self.k_norm(k)
        q = apply_rotary_emb(
            q,
            cos,
            sin,
            cu_seqlens=cu_seqlens_q,
            max_seqlen=max_seqlen_q,
            inplace=True,
        )
        k = apply_rotary_emb(
            k,
            cos,
            sin,
            cu_seqlens=cu_seqlens_k,
            max_seqlen=max_seqlen_k,
            inplace=True,
        )
        out = flash_attn_varlen_func(
            q=q,
            k=k,
            v=v,
            cu_seqlens_q=cu_seqlens_q,
            cu_seqlens_k=cu_seqlens_k,
            max_seqlen_q=max_seqlen_q,
            max_seqlen_k=max_seqlen_k,
            causal=self.causal,
            softmax_scale=self.scale,
        )
        out = out.view(total_nnz, dim)
        return self.to_out(out)


class TransformerBlock(nn.Module):
    def __init__(
        self,
        dim,
        heads,
        dim_head,
        intermediate_size,
        activation="relu_sq",
        causal: bool = True,
    ):
        super().__init__()
        self.attn_norm = RMSNorm(dim)
        self.attn = Attention(dim, heads=heads, dim_head=dim_head, causal=causal)
        self.mlp_norm = RMSNorm(dim)
        self.mlp = MLP(
            in_features=dim,
            hidden_features=intermediate_size,
            activation=activation,
        )

    def forward(
        self,
        hidden_states,
        cos: torch.Tensor,
        sin: torch.Tensor,
        cu_seqlens_q: Optional[torch.Tensor] = None,
        cu_seqlens_k: Optional[torch.Tensor] = None,
        max_seqlen_q: Optional[int] = None,
        max_seqlen_k: Optional[int] = None,
    ):
        hidden_states = hidden_states + self.attn(
            self.attn_norm(hidden_states),
            cos=cos,
            sin=sin,
            cu_seqlens_q=cu_seqlens_q,
            cu_seqlens_k=cu_seqlens_k,
            max_seqlen_q=max_seqlen_q,
            max_seqlen_k=max_seqlen_k,
        )
        hidden_states = hidden_states + self.mlp(self.mlp_norm(hidden_states))
        return hidden_states


class Transformer(nn.Module):
    def __init__(
        self,
        dim,
        depth,
        heads,
        dim_head,
        intermediate_size,
        activation="relu_sq",
        causal: bool = True,
    ):
        super().__init__()
        self.layers = nn.ModuleList(
            [
                TransformerBlock(
                    dim,
                    heads,
                    dim_head,
                    intermediate_size,
                    activation=activation,
                    causal=causal,
                )
                for _ in range(depth)
            ]
        )

    def forward(
        self,
        hidden_states: torch.Tensor,
        cos: torch.Tensor,
        sin: torch.Tensor,
        cu_seqlens_q: Optional[torch.Tensor] = None,
        cu_seqlens_k: Optional[torch.Tensor] = None,
        max_seqlen_q: Optional[int] = None,
        max_seqlen_k: Optional[int] = None,
    ):
        for layer in self.layers:
            hidden_states = layer(
                hidden_states,
                cu_seqlens_q=cu_seqlens_q,
                cu_seqlens_k=cu_seqlens_k,
                max_seqlen_q=max_seqlen_q,
                max_seqlen_k=max_seqlen_k,
                cos=cos,
                sin=sin,
            )
        return hidden_states


class RotaryEmbedding(nn.Module):
    def __init__(self, head_dim, base=10000):
        super().__init__()
        inv_freq = 1.0 / (
            base ** (torch.arange(0, head_dim, 2, dtype=torch.float32) / head_dim)
        )
        self.register_buffer("inv_freq", inv_freq, persistent=False)
        self.cos_sin_cache = None

    def get_freqs_non_repeated(self, max_seq_len: int):
        seq = torch.arange(
            max_seq_len, device=self.inv_freq.device, dtype=self.inv_freq.dtype
        )
        freqs = torch.outer(seq, self.inv_freq)
        return freqs

    def get_cos_sin(self, max_seq_len: int):
        freqs = self.get_freqs_non_repeated(max_seq_len)
        cos = torch.cos(freqs)
        sin = torch.sin(freqs)
        return cos, sin

    def forward(self, x, max_seq_len):
        if (
            self.cos_sin_cache is not None
            and self.cos_sin_cache[0].shape[0] >= max_seq_len
            and self.cos_sin_cache[0].device == x.device
            and self.cos_sin_cache[0].dtype == x.dtype
        ):
            cos, sin = self.cos_sin_cache
            return cos[:max_seq_len, :], sin[:max_seq_len, :]
        with torch.autocast(device_type=x.device.type, enabled=False):  # Force float32
            cos, sin = self.get_cos_sin(max_seq_len)
        self.cos_sin_cache = (cos.to(dtype=x.dtype), sin.to(dtype=x.dtype))
        return self.cos_sin_cache


class Model(nn.Module):
    def __init__(
        self,
        num_classes,
        dim,
        depth,
        heads,
        dim_head,
        intermediate_size,
        causal: bool = True,
    ):
        super().__init__()
        self.rotary_emb = RotaryEmbedding(dim_head)
        self.token_emb = TokenEmbeddings(num_classes, dim)
        self.transformer = Transformer(
            dim,
            depth,
            heads,
            dim_head,
            intermediate_size,
            causal=causal,
        )

    def forward(
        self,
        input_ids: torch.Tensor,
        cu_seqlens_q: Optional[torch.Tensor] = None,
        cu_seqlens_k: Optional[torch.Tensor] = None,
        max_seqlen_q: Optional[int] = None,
        max_seqlen_k: Optional[int] = None,
    ):
        x = self.token_emb(input_ids)
        cos, sin = self.rotary_emb(x, max_seqlen_q)

        return self.transformer(
            x,
            cos=cos,
            sin=sin,
            cu_seqlens_q=cu_seqlens_q,
            cu_seqlens_k=cu_seqlens_k,
            max_seqlen_q=max_seqlen_q,
            max_seqlen_k=max_seqlen_k,
        )


class Bert(nn.Module):
    def __init__(
        self,
        num_classes,
        dim,
        depth,
        heads,
        dim_head,
        intermediate_size,
        ignore_idx: int = -100,
        reduction: str = "mean",
        chunk_size: int = 4096,
        tuned: bool = True,
        causal: bool = False,
    ):
        super().__init__()

        self.model = Model(
            num_classes=num_classes,
            dim=dim,
            depth=depth,
            heads=heads,
            dim_head=dim_head,
            intermediate_size=intermediate_size,
            causal=causal,
        )

        self.head = PredictionHead(dim)

        self.decoder = nn.Linear(dim, num_classes, bias=False)
        self.decoder.weight = self.model.token_emb.tok_embeddings.weight

        self.ignore_idx = ignore_idx
        self.reduction = reduction
        self.chunk_size = chunk_size
        self.tuned = tuned

    def forward(
        self,
        input_ids: torch.Tensor,
        labels: torch.Tensor,
        cu_seq_lens_q: Optional[torch.Tensor] = None,
        cu_seq_lens_k: Optional[torch.Tensor] = None,
        max_length_q: Optional[int] = None,
        max_length_k: Optional[int] = None,
    ):
        output = self.model(
            input_ids=input_ids,
            cu_seqlens_q=cu_seq_lens_q,
            cu_seqlens_k=cu_seq_lens_k,
            max_seqlen_q=max_length_q,
            max_seqlen_k=max_length_k,
        )

        hidden_states = self.head(output)

        loss = chunked_linear_cross_entropy(
            hidden_states,
            self.decoder.weight,
            labels,
            chunk_size=self.chunk_size,
            ignore_index=self.ignore_idx,
            reduction=self.reduction,
            tuned=self.tuned,
        )

        return loss
```
This is the current iteration of the code. On a single GH200 it gets around 350k-550k TPS depending on the configuration I tested.

If I could get 300k TPS with flex attention, I would be more than happy and swap to that if it supports attention masks. I would use SDPA, but that seemed to be way too slow.

For some reason, this model is tanking TPS under DDP. Still trying to figure that out. Might be a recompile in Quack kernels.

**Comment by drisspg** [comment]
Sorry for late reponse, ill take a look
